src/invidious/yt_backend/url_sanitizer.cr

```diff
@@ -0,0 +1,121 @@
+      "playlist",                           # Unnamed playlist (id,id,id,...) (embed-only?)
+      "t", "time_continue", "start", "end", # Timestamp
+      "lc",                                 # Highlighted comment (watch page only)
+    ],
+  }
+
+  # Returns wether the given string is an ASCII word. This is the same as
+  # running the following regex in US-ASCII locale: /^[\w-]+$/
+  private def ascii_word?(str : String) : Bool
+    if str.bytesize == str.size
+      str.each_byte do |byte|
+        next if 'a'.ord <= byte <= 'z'.ord
+        next if 'A'.ord <= byte <= 'Z'.ord
+        next if '0'.ord <= byte <= '9'.ord
+        next if byte == '-'.ord || byte == '_'.ord
+
+        return false
+      end
+
+      return true
+    else
+      return false
+    end
+  end
+
+  # Return which kind of parameters are allowed based on the
+  # first path component (breadcrumb 0).
+  private def determine_allowed(path_root : String)
+    case path_root
+    when "watch", "w", "v", "embed", "e", "shorts", "clip"
+      return :watch
+    when .starts_with?("@"), "c", "channel", "user", "profile", "attribution_link"
+      return :channel
+    when "playlist", "mix"
+      return :playlist
+    when "results", "search"
+      return :search
+    else # hashtag, post, trending, brand URLs, etc..
+      return nil
+    end
+  end
+
+  # Create a new URI::Param containing only the allowed parameters
+  private def copy_params(unsafe_params : URI::Params, allowed_type) : URI::Params
+    new_params = URI::Params.new
+
+    ALLOWED_QUERY_PARAMS[allowed_type].each do |name|
+      if unsafe_params[name]?
+        # Only copy the last parameter, in case there is more than one
+        new_params[name] = unsafe_params.fetch_all(name)[-1]
+      end
+    end
+
+    return new_params
+  end
+
+  # Transform any user-supplied youtube URL into something we can trust
+  # and use across the code.
+  def process(str : String) : URI
+    # Because URI follows RFC3986 specifications, URL without a scheme
+    # will be parsed as a relative path. So we have to add a scheme ourselves.
+    str = "https://#{str}" if !str.starts_with?(/https?:\/\//)
+
+    unsafe_uri = URI.parse(str)
+    new_uri = URI.new(path: "/")
+
+    # Redirect to homepage for bogus URLs
+    return new_uri if (unsafe_uri.host.nil? || unsafe_uri.path.nil?)
+
+    breadcrumbs = unsafe_uri.path
+      .split('/', remove_empty: true)
+      .compact_map do |bc|
+        # Exclude attempts at path trasversal
+        next if bc == "." || bc == ".."
+
+        # Non-alnum characters are unlikely in a genuine URL
+        next if !ascii_word?(bc)
+
+        bc
+      end
+
+    # If nothing remains, it's either a legit URL to the homepage
+    # (who does that!?) or because we filtered some junk earlier.
+    return new_uri if breadcrumbs.empty?
+
+    # Replace the original query parameters with the sanitized ones
+    case unsafe_uri.host.not_nil!
```

Comment:
Could you rewrite this section to not use `not_nil!`? I'd like to slowly remove Invidious's dependence on it since it is generally frowned upon (see https://crystal-lang.org/api/1.10.1/Object.html#not_nil!-instance-method) and would allow for easier diagnosis of the various NilAssertionError seen across Invidious when designed right.

Reply:
If I do the following
```crystal
    # Replace the original query parameters with the sanitized ones
    case unsafe_uri.host
    when .ends_with?("youtube.com")
      # ...
    end
```

then the compiler complains that `Nil` doesn't have a `#ends_with?` method, even if I checked that it wasn't nil beforehand:

```crystal
# Redirect to homepage for bogus URLs
return new_uri if (unsafe_uri.host.nil? || unsafe_uri.path.nil?)
```

Reply:
I can make an intermediary variable (before the `if` check) if you want!
That's the only other way to make the compiler happy.

Reply:
Yes, please do. Intermediate variables seems to be the intended way to handle nilable properties within Crystal. There's an interesting discussion here about it https://forum.crystal-lang.org/t/is-casting-the-idiomatic-way-of-dealing-with-nil/1693